### PR TITLE
Introduce cluster level EvictionStrategy

### PR DESCRIPTION
**Pull request by acardace**
use LiveMigrate as a default value and a couple of unit tests.

Signed-off-by: Antonio Cardace <acardace@redhat.com>

**Reviewer Checklist**

> Reviewers are supposed to review the PR for every aspect below one by one. To check an item means the PR is either "OK" or "Not Applicable" in terms of that item. All items are supposed to be checked before merging a PR. 

- [ ] PR Message
- [ ] Commit Messages
- [ ] How to test
- [ ] Unit Tests
- [ ] Functional Tests
- [ ] User Documentation
- [ ] Developer Documentation
- [ ] Upgrade Scenario
- [ ] Uninstallation Scenario
- [ ] Backward Compatibility
- [ ] Troubleshooting Friendly
  
**Release note**:

```release-note
Introduce cluster level EvictionStrategy field.
```

### CODE REVIEW

**Comment by acardace** [comment]
@tiraboschi can anyone take a look at this?

**Comment by tiraboschi** [comment]
@acardace should this match Kubevirt values?
https://github.com/kubevirt/kubevirt/blob/16ca54fc6e4382a78ff315cc873cc692cdd1d62d/staging/src/kubevirt.io/api/core/v1/types.go#L1875-L1879

**Comment by tiraboschi** [comment]
> @acardace should this match Kubevirt values? https://github.com/kubevirt/kubevirt/blob/16ca54fc6e4382a78ff315cc873cc692cdd1d62d/staging/src/kubevirt.io/api/core/v1/types.go#L1875-L1879

Yes, the issue with functional tests comes from this.
We are getting
```
Internal error occurred: failed calling webhook "virtualmachineinstances-create-validator.kubevirt.io": failed to call webhook: Post "[https://virt-api.kubevirt-hyperconverged.svc:443/virtualmachineinstances-validate-create?timeout=10s](https://virt-api.kubevirt-hyperconverged.svc/virtualmachineinstances-validate-create?timeout=10s)": EOF
```
and indeed in https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/kubevirt_hyperconverged-cluster-operator/2071/pull-ci-kubevirt-hyperconverged-cluster-operator-main-hco-e2e-image-index-aws/1567189626756534272/artifacts/hco-e2e-image-index-aws/e2e-test/artifacts/kubevirt-must-gather/quay-io-kubevirt-must-gather-sha256-675cac0b2d2ef3ef3dcce5e16d75cae174a16074499c3253fda184c465ce2746/namespaces/kubevirt-hyperconverged/pods/virt-api-7fc64dfb68-ndn8x/virt-api/virt-api/logs/current.log

we see:
```
2022-09-06T18:57:29.243345170Z {"component":"virt-api","level":"info","msg":"http: panic serving 10.128.0.1:38264: runtime error: invalid memory address or nil pointer dereference\ngoroutine 3597 [running]:\nnet/http.(*conn).serve.func1()\n\tGOROOT/src/net/http/server.go:1802 +0xb9\npanic({0x1d2dc60, 0x382a5e0})\n\tGOROOT/src/runtime/panic.go:1047 +0x266\nkubevirt.io/kubevirt/pkg/virt-api/webhooks/validating-webhook/admitters.validateLiveMigration(0xc0088685a0, 0xc006802798, 0x0)\n\tpkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:779 +0x145\nkubevirt.io/kubevirt/pkg/virt-api/webhooks/validating-webhook/admitters.ValidateVirtualMachineInstanceSpec(0x1fd174e, 0xc006802798, 0x0)\n\tpkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:208 +0x1d4b\nkubevirt.io/kubevirt/pkg/virt-api/webhooks/validating-webhook/admitters.(*VMICreateAdmitter).Admit(0xc0004b11c0, 0xc008779c50)\n\tpkg/virt-api/webhooks/validating-webhook/admitters/vmi-create-admitter.go:100 +0x10c\nkubevirt.io/kubevirt/pkg/util/webhooks/validating-webhooks.Serve({0x2507758, 0xc008716ee0}, 0xc008716f20, {0x24ccc60, 0xc0004b11c0})\n\tpkg/util/webhooks/validating-webhooks/validating-webhook.go:73 +0xc9\nkubevirt.io/kubevirt/pkg/virt-api/webhooks/validating-webhook.ServeVMICreate(...)\n\tpkg/virt-api/webhooks/validating-webhook/validating-webhook.go:34\nkubevirt.io/kubevirt/pkg/virt-api.(*virtAPIApp).registerValidatingWebhooks.func1({0x2507758, 0xc008716ee0}, 0x0)\n\tpkg/virt-api/api.go:773 +0x7a\nnet/http.HandlerFunc.ServeHTTP(0x0, {0x2507758, 0xc008716ee0}, 0xc008716ee0)\n\tGOROOT/src/net/http/server.go:2047 +0x2f\nnet/http.(*ServeMux).ServeHTTP(0xc0086e379e, {0x2507758, 0xc008716ee0}, 0xc00846b600)\n\tGOROOT/src/net/http/server.go:2425 +0x149\nnet/http.serverHandler.ServeHTTP({0x24f8900}, {0x2507758, 0xc008716ee0}, 0xc00846b600)\n\tGOROOT/src/net/http/server.go:2879 +0x43b\nnet/http.(*conn).serve(0xc0069e2500, {0x250a218, 0xc00261af00})\n\tGOROOT/src/net/http/server.go:1930 +0xb08\ncreated by net/http.(*Server).Serve\n\tGOROOT/src/net/http/server.go:3034 +0x4e8\n","pos":"server.go:3160","timestamp":"2022-09-06T18:57:29.243292Z"}
```

So yes, we are passing a bad value here but `"virtualmachineinstances-create-validator.kubevirt.io"` should definitively be more resilient on bad configurations.

File: docs/cluster-configuration.md [hunk 1]
```diff
@@ -750,6 +750,23 @@ Please correctly consider the implications of this option before setting it.
...
+`UninstallStrategy` defines how to proceed on uninstall when workloads (VirtualMachines, DataVolumes) still exist:
+
+`EvictionStrategy` defines at the cluster level if the VirtualMachineInstance should be
+migrated instead of shut-off in case of a node drain. If the VirtualMachineInstance specific
+field is set it overrides the cluster level one.
+Possible values:
+
+- `EvictionStrategyNone` no eviction strategy at cluster level.
+- `EvictionStrategyLiveMigrate` ???
+- `EvictionStrategyExternal` ???
```
Comment (tiraboschi) [inline docs/cluster-configuration.md:765]: Can you please document the behaviour here?

Comment (acardace) [reply docs/cluster-configuration.md]: done, thank you.

**Comment by acardace** [comment]
@tiraboschi do you understand why all the index tests are failing?

**Comment by tiraboschi** [comment]
> @tiraboschi do you understand why all the index tests are failing?

They are failing here on the test for the default value:
```
++ oc get hco -n kubevirt-hyperconverged kubevirt-hyperconverged -o 'jsonpath={.spec.evictionStrategy}'
+ EVICTION_STRATEGY=LiveMigrate
+ [[ EvictionStrategyLiveMigrate != \L\i\v\e\M\i\g\r\a\t\e ]]
+ echo 'Failed checking CR defaults for evictionStrategy'
```

please fix the expected value on line 33 of hack/check_defaults.sh

File: docs/cluster-configuration.md [hunk 2]
```diff
@@ -742,14 +742,31 @@ can be applied.
...
+`EvictionStrategy` defines at the cluster level if the VirtualMachineInstance should be
+migrated instead of shut-off in case of a node drain. If the VirtualMachineInstance specific
+field is set it overrides the cluster level one.
+Possible values:
+
+- `EvictionStrategyNone` no eviction strategy at cluster level.
+- `EvictionStrategyLiveMigrate` migrate the VM on eviction.
+- `EvictionStrategyExternal` block eviction and notify an external controller.
+
+`EvictionStrategyLiveMigrate` is the default behaviour.
```
Comment (tiraboschi) [inline docs/cluster-configuration.md:767]:
```suggestion
- `None` no eviction strategy at cluster level.
- `LiveMigrate` migrate the VM on eviction.
- `External` block eviction and notify an external controller.

`LiveMigrate` is the default behaviour.
```

**Review by tiraboschi — APPROVED** [review]

**Comment by acardace** [comment]
@tiraboschi can you take a look? I've just rebased.

**Comment by tiraboschi** [comment]
this is supposed to be only in the future release-1.9
/hold

**Comment by acardace** [comment]
@tiraboschi can you take a look? I've rebased and merged all conflicts.

File: docs/cluster-configuration.md [hunk 3]
```diff
@@ -780,14 +780,31 @@ can be applied.
...
 - `RemoveWorkloads` will cause all the workloads to be cascading deleted on uninstallation.
 **WARNING**: please notice that RemoveWorkloads will cause your workloads to be deleted as soon as this CR will be, even accidentally, deleted.
 Please correctly consider the implications of this option before setting it.
 
 `BlockUninstallIfWorkloadsExist` is the default behaviour.
 
+
+## Cluster level EvictionStrategy
+
+`UninstallStrategy` defines how to proceed on uninstall when workloads (VirtualMachines, DataVolumes) still exist:
```
Comment (nunnatsa) [inline docs/cluster-configuration.md:794]: Is it blong here? Also, this is a duplication of line 782.

Comment (acardace) [reply docs/cluster-configuration.md]: Removed it.

**Review by tiraboschi — APPROVED** [review]

**Pull request merged**